core/src/test/java/kafka/server/share/SharePartitionTest.java

```diff
@@ -6190,6 +6198,289 @@ public void testFindLastOffsetAcknowledgedWhenGapAtBeginning() {
+        // In this situation, the last batch's offset state tracking is initialized. This is handled
+        // correctly because the client will send individual offset acknowledgements, which require offset
+        // state tracking anyway. While this last scenario is unlikely in practice (as a batch's reported
+        // last offset should remain correct even after compaction), the test verifies its proper handling.
+        fetchAcquiredRecords(sharePartition, records, 59);
+        assertEquals(64, sharePartition.nextFetchOffset());
+    }
+
+    /**
+     * This test verifies that cached batches which are no longer returned in fetch responses (starting
+     * from the fetchOffset) are correctly archived. Archiving these batches is crucial for the SPSO
+     * and the next fetch offset to advance. Without archiving, these offsets would be stuck, as the
+     * cached batches would remain available.
+     * <p>
+     * This scenario can occur with compacted topics when entire batches, previously held in the cache,
+     * are removed from the log at the offset where reading occurs.
+     */
+    @Test
+    public void testAcquireWhenBatchesRemovedForFetchOffset() {
+        SharePartition sharePartition = SharePartitionBuilder.builder()
+            .withState(SharePartitionState.ACTIVE)
+            .build();
+
+        fetchAcquiredRecords(sharePartition, memoryRecords(5, 0), 5);
+        fetchAcquiredRecords(sharePartition, memoryRecords(5, 5), 5);
+        fetchAcquiredRecords(sharePartition, memoryRecords(15, 10), 15);
+        // Release the batches in the cache.
+        sharePartition.releaseAcquiredRecords(MEMBER_ID);
+        // Validate cache has 3 entries.
+        assertEquals(3, sharePartition.cachedState().size());
+
+        // Compact second batch and remove first batch from the fetch response.
+        ByteBuffer buffer = ByteBuffer.allocate(4096);
+        try (MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, Compression.NONE,
+            TimestampType.CREATE_TIME, 5, 2)) {
+            // Append only 4 records for 5th offset batch starting from offset 6.
+            memoryRecords(4, 6).records().forEach(builder::append);
+        }
+        buffer.flip();
+        MemoryRecords records = MemoryRecords.readableRecords(buffer);
+
+        // Only second batch should be acquired and first batch offsets should be archived. Send
+        // fetchOffset as 0.
+        fetchAcquiredRecords(sharePartition, records, 0, 0, 5);
+        assertEquals(10, sharePartition.nextFetchOffset());
+        // Though the next fetch offset is moved but start offset should remain the same as we acquire
+        // just marks the offsets archived. The start offset shall move correctly once any records are acknowledged.
+        assertEquals(0, sharePartition.startOffset());
+
+        // Even invoking the release API shall update the cache and move the start offset.
```

Comment:
I think this is better as "Releasing acquired records updates the cache and moves the start offset."

Reply:
Done.